dali/pipeline/data/buffer.h

```diff
@@ -471,12 +474,8 @@ class DLL_PUBLIC Buffer {
     shares_data_ = data_ != nullptr;
     num_bytes_ = bytes;
     pinned_ = pinned;
-    // setting the allocation, get the device
-    if ((std::is_same<Backend, GPUBackend>::value || pinned_) && device_ == CPU_ONLY_DEVICE_ID) {
-      CUDA_CALL(cudaGetDevice(&device_));
-    } else {
-      device_ = CPU_ONLY_DEVICE_ID;
-    }
+    device_ = device_id;
+    set_order(order);
```

Comment:
My understanding of `set_order` (for two streams) is that I had a buffer that may be somehow used by the work scheduled on stream1, now I want to repupose the same memory for the work on stream2 so anything scheduled from then on on the stream2 will wait for the work hitherto scheduled on the stream1 to complete. But if I replace the underlying memory itself? Is there any connection between the two? (Esp. given the free_storage) at the top. Shouldn't it be `set_order(order, false)`?

Reply:
Actually, it should probably be moved just after the `free_storage`, as is already the case in ShareData for current TensorList implementation. That would make sense - `free_storage` might not free data (if we ware sharing), so maybe it's good idea to synchronize before we do the actual swap.